crates/fuzzing/src/oracles/diff_wasmi.rs

```diff
@@ -141,35 +129,36 @@ impl DiffInstance for WasmiInstance {
         arguments: &[DiffValue],
         result_tys: &[DiffValueType],
     ) -> Result<Option<Vec<DiffValue>>> {
-        let function = match self
+        let function = self
             .instance
             .get_export(&self.store, function_name)
-            .unwrap()
-        {
-            wasmi::Extern::Func(f) => f,
-            _ => unreachable!(),
-        };
+            .and_then(wasmi::Extern::into_func)
+            .unwrap();
         let arguments: Vec<_> = arguments.iter().map(|x| x.into()).collect();
         let mut results = vec![wasmi::core::Value::I32(0); result_tys.len()];
         function
             .call(&mut self.store, &arguments, &mut results)
             .context("wasmi function trap")?;
-        Ok(Some(results.into_iter().map(|x| x.into()).collect()))
+        Ok(Some(results.into_iter().map(Into::into).collect()))
     }
 
     fn get_global(&mut self, name: &str, _ty: DiffValueType) -> Option<DiffValue> {
-        match self.instance.get_export(&self.store, name).unwrap() {
-            wasmi::Extern::Global(g) => Some(g.get(&self.store).into()),
-            _ => unreachable!(),
-        }
+        self.instance
+            .get_export(&self.store, name)
+            .and_then(wasmi::Extern::into_global)
+            .map(|global| DiffValue::from(global.get(&self.store)))
+            .unwrap()
+            .into()
```

Comment:
Does this work? I like having a separate `unwrap` for each failure case so they generate different panic messages. (`rustfmt` will probably disagree with my formatting, I'm just typing in the GitHub web UI here.)

```suggestion
        Some(self.instance
            .get_export(&self.store, name)
            .unwrap()
            .into_global()
            .unwrap()
            .get(&self.store)
            .into())
```

If that works, I'd do the same in `get_memory` and at the top of `evaluate`.

Reply:
haha this actually looks even cleaner to me somehow ... :sweat_smile: will apply those changes to both sites